KSPCommunityFixes/Performance/FastLoader.cs

```diff
@@ -557,10 +731,15 @@ static IEnumerator FastAssetLoader(List<ConfigFileType> configFileTypes)
 
             // start model loading
             gdb.progressFraction = 0.75f;
+            KSPCFFastLoaderReport.wTextureLoading.Stop();
+            KSPCFFastLoaderReport.wModelLoading.Start();
             gdb.progressTitle = "Loading model assets...";
             yield return null;
 
             // call non-stock model loaders
+            modelsByUrl = new Dictionary<string, GameObject>(allModelFiles.Count);
```

Comment:
Why not
```
modelsByUrl = new Dictionary<string, GameObject>(allModelFiles.Count, StringComparer.OrdinalIgnoreCase);
modelsByDirectoryUrl = new Dictionary<string, GameObject>(allModelFiles.Count, StringComparer.OrdinalIgnoreCase);
```
And the same for texturesByUrl?
It should have the same semantics as your existing equality logic, but probably will be faster based on your comment elsewhere about frequently falling through.
It might be slightly slower in the case where it's an exact match,  but not drastically - and it should eliminate fallback to the O(n) case for case insensitive matches. I don't know if you still need the fallback (I saw in some cases that you might have to check a List because you could just directly modify it in stock KSP code for one of these optimizations), but thought I'd mention it.
https://learn.microsoft.com/en-us/dotnet/api/system.stringcomparer.ordinalignorecase?view=net-8.0
https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.dictionary-2.-ctor?view=net-8.0#system-collections-generic-dictionary-2-ctor(system-int32-system-collections-generic-iequalitycomparer((-0)))

Reply:
Model search is case sensitive, but yeah that could make sense for textures.

TBH, occurrences of the methods being called with mismatching casing were pretty rare in my tests (less than 1%), and for the few cases where it happens, the requested casing will be added as duplicate entry, so the overhead will only occur on the first call for a specific casing. The case insensitive comparer will still induce quite a bit of overhead, so I'm not so sure that would be a net gain overall.

Falling back to a linear search on the original stock lists is necessary in any case (:P), as (a few) models/textures are also loaded from asset bundles and because there nothing preventing external code from modifying the lists directly. That still leave a hole if someone decide to *remove* models/textures from them, but that feel unlikely enough.

On a side note, upon further review, we should probably also patch `GetTextureInfoIn()` and `GetModel()`, as those are commonly used too, including during part compilation.

Reply:
Fair - I assumed it was significantly more frequent based on this comment on the issue (which I happened across by chance when being curious about what the latest version you just published held :)).

"The dictionary approach is however not as beneficial here, because those methods allow to get a texture by its case-insensitive url, so we have to fallback to iterating on the list with a case insensitive string comparison if the requested url isn't correctly cased. The net result was still beneficial in my tests, but not as much as I hoped **due to that slow path being taken relatively often.**"

InvariantCultureIgnoreCase and CurrentCultureIgnoreCase are fairly expensive; I'd be surprised (but would believe you) if OrdinalIgnoreCase was particularly expensive though. Then again, my knowledge may be more relevant for newer dotnet versions? (ex: https://devblogs.microsoft.com/dotnet/performance-improvements-in-net-8/).

But fair, you're trading an initial miss + a bit of memory in an uncommon case, to save a bit of arithmetic / bit-twiddling + possibly other overheads on every case - I'm not a KSP modding expert, your intuition is likely better than mine in this area and you're in a position to measure it for sure :).

Pity you can't get rid of that silly list - I was gonna suggest "well if you can modify, why not binary search" - but that won't work if code makes additions without being aware of the need to insert at the right position.

Reply:
Yeah, unfortunately, the mono-derived BCL and JIT compiler used in KSP/Unity is missing all the juicy improvements from more or less the last 10 years (and in many cases, perform just plain worse than the good old .Net framework).

This being said, the case-insensitive dictionary still make sense and is a trivial change, so I will try and check what the profiling results have to say.

Reply:
Ah, right... I haven't touched Unity in about half a decade. My frame of reference for "old dotnet performance" is .net framework... I forgot that Unity has its own fork of Mono.
And I'm assuming Unity's version of  Mono is just different enough / just tightly coupled enough that we can't e.g. just drop in a modern version of the dotnet runtime either... Though it might be fun to try sometime :smile: